### PR TITLE
Batch Normalization typecheck

### PR DESCRIPTION
**Pull request by delta2323**
* Add typecheck to BatchNormalization

Related to #123

### CODE REVIEW

File: chainer/functions/batch_normalization.py [hunk 1]
```diff
@@ -91,6 +92,7 @@ class BatchNormalization(function.Function):
     gradient_names = ('ggamma', 'gbeta')
 
     def __init__(self, size, decay=0.9, eps=1e-5):
+        self.size = size
```
Comment (beam2d) [inline chainer/functions/batch_normalization.py]: How about forcing `self.size` to be a tuple? E.g.: `self.size = size if isinstance(size, tuple) else (size,)`?

File: chainer/functions/batch_normalization.py [hunk 2]
```diff
@@ -130,6 +132,35 @@ def __call__(self, x, test=False, finetune=False):
         self.is_finetune = finetune
         return function.Function.__call__(self, x)
 
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+        x_type, = in_types
+
+        if isinstance(self.size, tuple):
+            feature_shape = type_check.Variable(self.size, 'feature_shape')
```
Comment (beam2d) [inline chainer/functions/batch_normalization.py]: How about making `self` as a variable (e.g. named `self_`) instead of feature_shape?

File: chainer/functions/batch_normalization.py [hunk 3]
```diff
@@ -130,6 +132,35 @@ def __call__(self, x, test=False, finetune=False):
...
+            feature_ndim = len(self.size)
+        else:
+            feature_shape = type_check.Variable((self.size,), 'feature_shape')
+            feature_ndim = 1
+
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim >= type_check.Variable(
+                feature_ndim, 'feature_ndim') + 1,
```
Comment (beam2d) [inline chainer/functions/batch_normalization.py]: Combined with the above comment, how about using `self_.size.__len__()` instead of the `feature_ndim` variable?

**Comment by beam2d** [comment]
I wrote some comments.

**Comment by beam2d** [comment]
LGTM :+1:

**Comment by beam2d** [comment]
Thank you!
